### PR TITLE
Implement a repository, a menu and a carousel

### PR DESCRIPTION
**Pull request by SavoGK**
- Implemented repository for movie and genre data
- Implemented a class movie category, this map a genre with a list of movies
- Implemented views home and load screen. Load screen wait for obtain the data and home show all the movies separated by genre
- Implemented a carousel to show the movies divided for genre

[Record Module 3.webm](https://github.com/SavoGK/TheMovieDB/assets/104601226/c81bc0de-ca0d-4da9-b9a6-7285bab24521)

Second Commit:
- Change the carousel for a grid and the genres are now in a drawer

gif module 3

### CODE REVIEW

File: lib/models/repository/genre_repository.dart [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import 'package:the_movie_db/utils/general_constants.dart';
+import '../genre.dart';
+import 'repository.dart';
+
+class GenreRepository implements Repository {
+  static const String jsonFilePath = 'data/genres.json';
+  static List<Genre> genres = [];
+
+  @override
+  Future<void> fetchData() async {
```
Comment (matiasilveiro-gl) [inline lib/models/repository/genre_repository.dart]: I wouldn't add this fetchData method as an interface method. You only need a get method, the implementation will take care of the where you get the data.
Same for the movies repo

File: lib/views/loading_screen.dart [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  );
+
+  void _loadNextScreen() async {
+    await MovieRepository().fetchData();
+    await GenreRepository().fetchData();
+    setState(
+      () {
+        _body = const Home();
+      },
+    );
```
Comment (matiasilveiro-gl) [inline lib/views/loading_screen.dart]: I'd use the `Navigator` instead of this

File: lib/widget/movie_carousel.dart [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+                      context,
+                      MaterialPageRoute(
+                        builder: (context) => MovieDeck(
+                          movie: movie,
+                          genres: genres,
+                        ),
+                      ),
+                    );
+                  },
+                  child: MoviePoster(movie: movie));
```
Comment (matiasilveiro-gl) [inline lib/widget/movie_carousel.dart]: Trailing comma

File: lib/widget/movie_carousel.dart [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+            itemBuilder: (context, index) {
+              final movie = movies[index];
+              return GestureDetector(
+                  onTap: () {
+                    Navigator.push(
+                      context,
+                      MaterialPageRoute(
+                        builder: (context) => MovieDeck(
+                          movie: movie,
+                          genres: genres,
```
Comment (matiasilveiro-gl) [inline lib/widget/movie_carousel.dart]: You're passing your genres across your entire widget tree. I wouldn't pass this list as a parameter, but get the genres again inside `MovieDeck` instead. You're even using caching in your repository!

File: lib/widget/movie_poster.dart [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+  @override
+  Widget build(BuildContext context) {
+    return Container(
+      margin: const EdgeInsets.all(EdgeInsetsConst.edgeTen),
+      child: Column(
+        children: [
+          Expanded(
+            child: Image.network(
+              ApiEndPoints.urlImages + movie.poster,
```
Comment (matiasilveiro-gl) [inline lib/widget/movie_poster.dart]: Let's create a getter method for the URL inside the `Movie` class for this!

File: lib/models/movie_category.dart [hunk 6]
```diff
@@ -0,0 +1,28 @@
+import 'genre.dart';
+import 'movie.dart';
+
+class MovieCategory {
+  final List<Genre> genres;
+  final List<Movie> movies;
+  static Map<int, List<Movie>> mapMovies = {};
+
+  MovieCategory({
```
Comment (FranciscoGaraventa) [inline lib/models/movie_category.dart]: The idea to this class is to separated every page that you can get from the api by a genre?

File: lib/views/home.dart [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        } else if (snapshot.hasError) {
+          return Text('$error ${snapshot.error}');
+        } else {
+          List<dynamic>? data = snapshot.data;
+          List<Movie> movies = data![positionMovies];
+          List<Genre> genres = data[positionGenres];
+          MovieCategory movieCategory = MovieCategory(
+            genres: genres,
+            movies: movies,
+          );
```
Comment (FranciscoGaraventa) [inline lib/views/home.dart]: Try to move this logic to another method

File: lib/views/home.dart [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+  static const int positionMovies = 0;
+  static const int positionGenres = 1;
+  static const String error = 'Error:';
+
+  @override
+  Widget build(BuildContext context) {
+    return FutureBuilder(
+      future: Future.wait(
+          [MovieRepository().getData(), GenreRepository().getData()]),
+      builder: (context, AsyncSnapshot<List<dynamic>> snapshot) {
```
Comment (FranciscoGaraventa) [inline lib/views/home.dart]: Add the type of context, and also you missed a trailing comma

File: lib/views/home.dart [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+class _HomeState extends State<Home> {
+  static const int positionMovies = 0;
+  static const int positionGenres = 1;
+  static const String error = 'Error:';
+
+  @override
+  Widget build(BuildContext context) {
+    return FutureBuilder(
+      future: Future.wait(
+          [MovieRepository().getData(), GenreRepository().getData()]),
```
Comment (FranciscoGaraventa) [inline lib/views/home.dart]: trailing comma

File: lib/views/loading_screen.dart [hunk 10]
```diff
@@ -0,0 +1,42 @@
+import 'package:flutter/material.dart';
+import 'home.dart';
+import '../models/repository/genre_repository.dart';
+import '../models/repository/movie_repository.dart';
+
+class LoadingScreen extends StatefulWidget {
```
Comment (FranciscoGaraventa) [inline lib/views/loading_screen.dart]: You need this loading screen? You can do the same on HomePage, using the waiting of the bloc

File: lib/widget/loader_image.dart [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+  static const String imageIconAdults = 'images/icon_adults.png';
+
+  @override
+  Widget build(BuildContext context) {
+    return Stack(
+      children: [
+        Image.network(
+          ApiEndPoints.urlImages + image,
+          fit: BoxFit.fill,
+          loadingBuilder: (BuildContext context, Widget child,
```
Comment (FranciscoGaraventa) [inline lib/widget/loader_image.dart]: trailing comma

File: lib/widget/movie_carousel.dart [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+                fontWeight: FontWeight.bold,
+                color: Colors.white),
+          ),
+        ),
+        SizedBox(
+          height: heightSizedBox,
+          child: ListView.builder(
+            scrollDirection: Axis.horizontal,
+            itemCount: movies.length,
+            itemBuilder: (context, index) {
```
Comment (FranciscoGaraventa) [inline lib/widget/movie_carousel.dart]: Add the type of the parameters, and also you missed a trailing comma

File: lib/widget/movie_poster.dart [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+              fit: BoxFit.fill,
+              loadingBuilder: (BuildContext context, Widget child,
+                  ImageChunkEvent? loadingProgress) {
```
Comment (FranciscoGaraventa) [inline lib/widget/movie_poster.dart]: trailing comma

File: lib/models/repository/genre_repository.dart (hunk 1 above)
Comment (SavoGK) [reply lib/models/repository/genre_repository.dart]: done

File: lib/views/loading_screen.dart (hunk 2 above)
Comment (SavoGK) [reply lib/views/loading_screen.dart]: I removed the view. I guess I can say it's done

File: lib/widget/movie_carousel.dart (hunk 3 above)
Comment (SavoGK) [reply lib/widget/movie_carousel.dart]: done

File: lib/widget/movie_poster.dart (hunk 5 above)
Comment (SavoGK) [reply lib/widget/movie_poster.dart]: done

File: lib/widget/movie_carousel.dart (hunk 4 above)
Comment (SavoGK) [reply lib/widget/movie_carousel.dart]: done

File: lib/views/home.dart (hunk 7 above)
Comment (SavoGK) [reply lib/views/home.dart]: done

File: lib/models/movie_category.dart (hunk 6 above)
Comment (SavoGK) [reply lib/models/movie_category.dart]: i removed it

File: lib/views/home.dart (hunk 8 above)
Comment (SavoGK) [reply lib/views/home.dart]: done

File: lib/views/home.dart (hunk 9 above)
Comment (SavoGK) [reply lib/views/home.dart]: done

File: lib/views/loading_screen.dart (hunk 10 above)
Comment (SavoGK) [reply lib/views/loading_screen.dart]: done

File: lib/widget/movie_poster.dart (hunk 13 above)
Comment (SavoGK) [reply lib/widget/movie_poster.dart]: done

File: lib/widget/movie_carousel.dart (hunk 12 above)
Comment (SavoGK) [reply lib/widget/movie_carousel.dart]: done

File: lib/widget/loader_image.dart (hunk 11 above)
Comment (SavoGK) [reply lib/widget/loader_image.dart]: done

File: lib/models/repository/genre_repository.dart [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+import 'package:flutter/services.dart';
+import 'package:the_movie_db/utils/general_constants.dart';
+import '../genre.dart';
+import 'repository.dart';
+
+class GenreRepository implements Repository {
+  static const String jsonFilePath = 'data/genres.json';
+  static List<Genre> genres = [];
+
+  @override
```
Comment (matiasilveiro-gl) [inline lib/models/repository/genre_repository.dart]: Remove the `@override`

File: lib/models/repository/movie_repository.dart [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+import 'package:flutter/services.dart';
+import 'repository.dart';
+import '../../utils/general_constants.dart';
+import '../movie.dart';
+
+class MovieRepository implements Repository {
+  static const String jsonFilePath = 'data/movies.json';
+  static List<Movie> movies = [];
+
+  @override
```
Comment (matiasilveiro-gl) [inline lib/models/repository/movie_repository.dart]: Remove the `@override`

File: lib/views/movie_deck.dart [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+  static const int position = 0;
+
+  const MovieDeck({
+    required this.movie,
+    super.key,
+  });
+
+  List<Genre> initializeData(
+      AsyncSnapshot<List<dynamic>> snapshot,
+      ) {
```
Comment (matiasilveiro-gl) [inline lib/views/movie_deck.dart]: Format this code. Also, I'd rename this method. You're not initializing anything, but filtering the genres!

File: lib/views/movie_deck.dart [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return FutureBuilder(
+      future: Future.wait(
+        [
+          GenreRepository().getData(),
+        ],
+      ),
```
Comment (matiasilveiro-gl) [inline lib/views/movie_deck.dart]: You can wait to a single future:
`future: GenreRepository().getData()`.
Also, we can add some best practices here. Please read [this article](https://api.flutter.dev/flutter/widgets/FutureBuilder-class.html#:~:text=a%20Future.-,Managing%20the%20future,-The%20future%20must) to refactor this widget.

File: lib/widget/drawer_genre.dart [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+                color: Colors.white),
+          ),
+          onTap: () {
+            List<Movie> moviesGenre = getMoviesByGenre(genre.id);
+            Navigator.push(
+              context,
+              MaterialPageRoute(
+                builder: (context) => MoviesGenre(
+                  genre: genre,
+                  movies: moviesGenre,
```
Comment (matiasilveiro-gl) [inline lib/widget/drawer_genre.dart]: You shouldn't have the movies here. You should fetch them in the `MoviesGenre` screen

File: lib/widget/drawer_genre.dart [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+  });
+
+  static const int mediaQueryConst = 230;
+  static const double heightSizeBox = 60;
+
+  List<Movie> getMoviesByGenre(int genreId) {
+    return movies.where((movie) => movie.genres.contains(genreId)).toList();
+  }
+
+  List<Widget> _getWidget(BuildContext context, List<Genre> genres) {
```
Comment (FranciscoGaraventa) [inline lib/widget/drawer_genre.dart:23]: trailing comma

**Review by FranciscoGaraventa — APPROVED** [review]

**Review by matiasilveiro-gl — APPROVED** [review]

**Review by MarcosCGlobant — APPROVED** [review]

**Pull request merged**